daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1105,7 +1105,7 @@ object SBuiltin {
           // update expression threw an exception. we're
           // now done.
           machine.clearCommit
-          machine.returnValue = SV.Unit
+          machine.returnValue = machine.userErrorMessage.map(msg => SText(msg)).getOrElse[SValue](SV.Unit)
```

Comment:
That looks dangerous. You are changing the return value even for `submitMustFail` which expects a `Unit`. This _might_ work in some cases because speedy does not change types but it definitely doesn’t work in all cases: Consider what happens if you compare the return value to `SV.Unit`later, that comparison should now fail whereas it succeeded before.
You can probably extend the parameters passed to `SbEndCommit` to differentiate between the two and only set the parameter if necessary.